File: graph/2DtoAdjacency.cpp
#include <iostream>
#include <vector>

using namespace std;

// Directions for the 8 possible moves (up, down, left, right, and diagonals)
int directions[8][2] = {
    {-1, -1}, {-1, 0}, {-1, 1},
    { 0, -1},          { 0, 1},
    { 1, -1}, { 1, 0}, { 1, 1}
};

// Function to check if the cell is valid and safe
bool is_safe(const vector<vector<char>>& M, int r, int c) {
    int rows = M.size();
    int cols = M[0].size();
    return (r >= 0 && r < rows && c >= 0 && c < cols && M[r][c] == '1');
}

// Function to convert 2D matrix to adjacency list
vector<vector<int>> convert_matrix_to_graph(const vector<vector<char>>& M) {
    int rows = M.size();
    int cols = M[0].size();
    int total_nodes = rows * cols;

    // Create an adjacency list
    vector<vector<int>> adj(total_nodes);

    for (int r = 0; r < rows; r++) {
        for (int c = 0; c < cols; c++) {
            if (M[r][c] == '1') {
                int idx = r * cols + c;
                adj[idx].push_back(idx); // Add self-loop for convenience

                for (int d = 0; d < 8; d++) {
                    int nr = r + directions[d][0];
                    int nc = c + directions[d][1];

                    if (is_safe(M, nr, nc)) {
                        int n_idx = nr * cols + nc;
                        adj[idx].push_back(n_idx); // Add neighbor
                    }
                }
            }
        }
    }

    return adj;
}

// DFS Utility Function
void dfs_util(int v, const vector<vector<int>>& adj, vector<bool>& visited) {
    visited[v] = true;

    for (int neighbor : adj[v]) {
        if (!visited[neighbor]) {
            dfs_util(neighbor, adj, visited);
        }
    }
}

// Function to count islands using DFS
int count_islands(const vector<vector<char>>& M) {
    int rows = M.size();
    int cols = M[0].size();
    int total_nodes = rows * cols;

    // Convert the matrix to an adjacency list
    vector<vector<int>> adj = convert_matrix_to_graph(M);

    // Initialize visited array
    vector<bool> visited(total_nodes, false);

    int count = 0;

    // Traverse all nodes
    for (int i = 0; i < total_nodes; i++) {
        if (!visited[i] && !adj[i].empty()) {
            dfs_util(i, adj, visited);
            count++; // Increment count for each connected component
        }
    }

    return count;
}

// Main function
int main() {
    // Example 2D matrix as a vector of vector<char>
    vector<vector<char>> M = {
        {'1', '1', '0', '0', '0'},
        {'0', '1', '0', '0', '1'},
        {'1', '0', '0', '1', '1'},
        {'0', '0', '0', '0', '0'},
        {'1', '0', '1', '1', '0'}
    };

    cout << "Number of Islands in M: " << count_islands(M) << endl;
    // Expected Output: 4

    return 0;
}

File: graph/2DtoAdjacency.py
from typing import List

def is_safe(M: List[List[str]], r: int, c: int, Row: int, Col: int):
    return (0 <= r < Row) and (0 <= c < Col) and (M[r][c] == '1')


def convert_matrix_to_graph(M: List[List[str]]) -> List[List[int]]:
    n_row = len(M)
    n_col = len(M[0])
    adj = [[] for _ in range(n_row * n_col)]

    directions = [[-1, -1], [-1, 0], [-1, 1],
                  [0, -1],          [ 0, 1],
                  [1, -1],  [1, 0], [1, 1]]
    
    for r in range(n_row):
        for c in range(n_col):
            if M[r][c] == '1':
                idx = r * n_col + c
                adj[idx].append(idx)
                for dr, dc in directions:
                    nr = r + dr
                    nc = c + dc
                    if is_safe(M, nr, nc, n_row, n_col):
                        n_idx = nr * n_col + nc
                        adj[idx].append(n_idx)
    return adj


def dfs(adj: list[list[int]]) -> int:
    cnt = 0
    
    n = len(adj)
    visited = [False] * n

    def dfs_util(v: int):
        visited[v] = True
        for neighbor in adj[v]:
            if not visited[neighbor]:
                dfs_util(neighbor)
    
    for i in range(n):
        if not visited[i] and len(adj[i]):
            dfs_util(i)
            cnt += 1
    
    return cnt


def count_islands(M: List[List[str]]) -> int:
    adj = convert_matrix_to_graph(M)
    print(adj)
    return dfs(adj)


if __name__ == "__main__":
    # Example Usage (Same as previous)
    M = [['1', '1', '0', '0', '0'],
         ['0', '1', '0', '0', '1'],
         ['1', '0', '0', '1', '1'],
         ['0', '0', '0', '0', '0'],
         ['1', '0', '1', '1', '0']]
    print("Number of Islands in M using BFS:", count_islands(M))
    # Expected Output: 4

    # Other test cases can be reused from the previous example